### PR TITLE
[dvsim] Maximize Scheduler performance, other improvements

### PR DESCRIPTION
**Pull request by sriyerg**
The 1st commit changes the way the `Scheduler` dispatches jobs - rather than iterate over each target (wait for all builds to finish, then start all runs, and so on), it iterates over a job's dependency. This enables runs of a particular build to be scheduled immediately once the build passes, rather than wait for all builds to complete. This allows the max parallelization slots to be fully utilized, allowing a good deal of speedup (speedup is more apparent when the max parallel slots is large). 

The 2nd commit adds weights to each target, so that the Scheduler can more smartly decide which job to schedule. The build and coverage targets are assigned a higher weight over simulation runs so that once all AES simulations finish for example, its coverage tasks do not need to wait until ALL other simulations in other DUTs are done - with weighted scheduling, they are dispatched sooner so that its coverage data is also available sooner. 

The 3rd commit enhances how the status is printed to the console, using an external progress bar tool called 'enlighten'. 

The last commit is a minor enhancement to `utils.clean_odirs` (outcome of profiling).

### CODE REVIEW

File: util/dvsim/Deploy.py
Comment (rswarbrick) [inline util/dvsim/Deploy.py]: This leaks an open file, I think. Also, catching `Exception` is always a bit dubious and I think your error message won't be formatted right (extra quotes on the filename and repr on an Exception object). Maybe something like this would be better?
```python3
        try:
            with open(self.log, "r", encoding="UTF-8") as f:
                lines = f.readlines()
        except OSError as e:
            msg = "Error opening file {!r}:\n{}".format(self.log, e)
            self.fail_msg += msg
            log.log(VERBOSE, msg)
            return False
```

File: util/dvsim/Deploy.py
Comment (rswarbrick) [inline util/dvsim/Deploy.py]: You don't actually need to do this clamping explicitly:
```
>>> l = list(range(10))
>>> l[4:30]
[4, 5, 6, 7, 8, 9]
```

File: util/dvsim/Deploy.py
Comment (rswarbrick) [inline util/dvsim/Deploy.py]: This pair of lines gets repeated a lot. Maybe it would be worth factoring out? (This might also allow you to get rid of some local variables, because you could just construct "msg" here as the argument to the logging function).

File: util/dvsim/Deploy.py
Comment (rswarbrick) [inline util/dvsim/Deploy.py]: This is neat, but it might be easier to remove the pass pattern that we match? That way, later calls to `_find_patterns` don't have to search for it.

File: util/dvsim/Deploy.py
Comment (rswarbrick) [inline util/dvsim/Deploy.py]: This sort of thing is why I don't like auto-formatters. It looks horrible! If we must use the auto-formatter, could you put the argument to `format` into a local variable so that the lines break in a slightly more sensible place?

The same comment goes for line 390 below.

File: util/dvsim/Deploy.py
Comment (rswarbrick) [inline util/dvsim/Deploy.py]: I think the terminology is normally "symbolic link" or "symlink".

File: util/dvsim/Deploy.py
Comment (rswarbrick) [inline util/dvsim/Deploy.py]: There are race conditions here. See https://stackoverflow.com/questions/8299386/modifying-a-symlink-in-python#27788271 for how to do this properly. Also (although this *would* be racy) I wonder whether we should also be checking that `dest` was a symlink before?

File: util/dvsim/Deploy.py
Comment (rswarbrick) [inline util/dvsim/Deploy.py]: Maybe it would make sense to move this logic into a helper function, since it gets used in a couple of places. Also, the usual way to do this is to try to delete, ignoring any exception, then check the directory no longer exists.

File: util/dvsim/Deploy.py
Comment (rswarbrick) [inline util/dvsim/Deploy.py]: This doesn't have quite the same semantics as it had: I think you probably don't want an error if `self.log` didn't happen to exist before. If you write the helper function suggested above, you could just use that.

File: util/dvsim/dvsim.py
Comment (rswarbrick) [inline util/dvsim/dvsim.py:78]: Do you want to handle the case when I give `--scratch-root=/root` with a nice error message here?

File: util/dvsim/utils.py
Comment (rswarbrick) [inline util/dvsim/utils.py:107]: Hmm, the problem with this is that it now always looks like a string. The advantage of `repr` is that the output is unambiguous: I think that adding quotes probably makes it more confusing, not less.

**Review by rswarbrick — DISMISSED** [review]
I've only read the first few commits so far. I've got a couple of suggested clean-ups, which would be good to apply. Maybe it makes sense to split them out into a separate PR so that we can get things merged in stages?

Also, the automated reformatting ("lint fixes") that you're doing are making the code much harder to read, not more clear. We don't currently have an OpenTitan policy that mandates a canonical formatting. Are you *really* sure it's a good idea? If you are, please can you split the auto-format commits out from your actual changes in future? It would make reviewing much easier.

EDIT: I was wrong. Apparently, we [do](https://docs.opentitan.org/doc/rm/python_coding_style) mandate yapf-generated formatting. Yuck. Please still split things up and (if possible) do gymnastics to try to avoid it making too much of a mess of the code.

File: util/dvsim/FlowCfg.py
Comment (sriyerg) [reply util/dvsim/FlowCfg.py:366]: Moved this commit to PR #5321.

File: util/dvsim/Scheduler.py
Comment (sriyerg) [reply util/dvsim/Scheduler.py:293]: Updated in #5321.

File: util/dvsim/utils.py
Comment (sriyerg) [reply util/dvsim/utils.py:107]: Removed.

**Review by sriyerg — COMMENTED** [review]
@rswarbrick, I have split this commit to a separate PR #5346 and I have addressed your comments there. Thanks.

File: util/dvsim/Deploy.py
Comment (rswarbrick) [reply util/dvsim/Deploy.py]: Looks good, thanks.

File: util/dvsim/Deploy.py
Comment (rswarbrick) [reply util/dvsim/Deploy.py]: Oh! You're right! I didn't know that, thanks.

File: util/dvsim/Deploy.py
Comment (rswarbrick) [reply util/dvsim/Deploy.py]: Well, you're right that we shouldn't run two copies of `dvsim` at the same time, but there's a standard atomic way to update a symlink, so I'd generally suggest just doing it properly.

File: util/dvsim/Deploy.py
Comment (rswarbrick) [reply util/dvsim/Deploy.py]: Well, it will run marginally faster (avoiding a system call) and the "try to do it and handle any problems" approach seems to be more idiomatic Python.

File: util/dvsim/Deploy.py
Comment (sriyerg) [reply util/dvsim/Deploy.py]: Done (in the other PR).

**Comment by sriyerg** [comment]
@rswarbrick, this PR a ready for review as well - this is a Scheduler-only refactor.

**Comment by rswarbrick** [comment]
Please could you rebase on top of your other merged changes?

**Comment by sriyerg** [comment]
> Please could you rebase on top of your other merged changes?

Yes, but I wanted to wait until #5417 landed so that I can resolve al conflicts together.

**Comment by sriyerg** [comment]
> > Please could you rebase on top of your other merged changes?
> 
> Yes, but I wanted to wait until #5417 landed so that I can resolve al conflicts together.

Rebased, PTAL, thanks!

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py]: As I understand it, a target gets added to `self._completed` once all its items are done. This set is then used to filter out targets from `self._scheduled`. Would it be easier to just remove the target from `self._scheduled` and get rid of `self._completed` entirely?

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py:134]: It might be cleaner to promote the nested function (`add_to_scheduled`) to be a proper method to match `_remove_from_scheduled` and then inline the top-level loop into the call site at `__init__`. Then there are no nested functions (always nice) and things look a bit more symmetrical.

If you don't want to do that, please make this a static method and also fix the docstring style.

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py:137]: I don't think there's any need for an `OrderedDict` here. Python 3.6 gives the ordering stability that I think you're using to plain `dict`s.

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py:153]: You can use `setdefault` to make this a bit simpler (and it will probably run faster too, not that we really care)
```python3
            for_tgt = scheduled.setdefault(item.target, [])
            for_cfg = for_tgt.setdefault(item.sim_cfg, [])
            if item not in for_cfg:
                for_cfg.append(item)
```

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py:168]: I think this sort of thing is much easier to read with some local variables to avoid the repeated lookups. Like this, for example:
```python3
        for_tgt = self._scheduled[item.target]
        for_cfg = for_tgt.get(item.sim_cfg)

        if for_cfg is not None:
            for_cfg.remove(item)
            if not for_cfg:
                del for_tgt[item.sim_cfg]
```
It also avoids `del foo[bar][baz]`, which I always have to think about. What do you think?

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py:171]: Can you add a paragraph to the docstring to explain that `curr_target` can be `None`?

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py:217]: Not for this PR, but this scheduler seems to be going most of the way to a full dependency DAG. I wonder whether it would be sensible to finish up that transition at some point: I think that some of the code in `_get_successors` and `_get_next_target` would actually get much simpler.

Also, we seem to be re-inventing Make, which is a pity. I wonder whether it would have been more sensible to use Make plus a job dispatch backend. (In a company I used to work for, we could use LSF's "`bmake`", which basically spawned everything through `bsub`. It was really bad manners to use on the queue, because it spawned hundreds of jobs, but it was lightning fast :-))

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py]: This is pretty hard to read because of the mess made by the auto-formatter. Maybe define a local variable for the second argument? Also, it's probably simpler not to include the length here: you can just check whether `self._queued[target]` is empty in the test of the while loop below. That also gets rid of the `try/except` block.

Finally, can you call this `num_slots_for_target` instead? We'll compute different values for different targets, and "per" makes it sound like this is something that applies across the board.

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py:379]: I think this logic should probably live in `_enqueue_successors`. That way, we can choose never to enqueue stuff that should be cancelled. This avoids us having to wait a polling loop per target if everything is broken.

Comment (rswarbrick) [reply util/dvsim/Scheduler.py:369]: Oh, I see. Good point!

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py:351]: Nit: Since `sum` can take a generator, you don't actually need the outer square brackets here.

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py:345]: Nit: many

File: util/dvsim/Scheduler.py
Comment (rswarbrick) [inline util/dvsim/Scheduler.py:359]: I'm probably just being a bit slow here, but I don't quite understand the weight computation. Would you mind adding a couple of comments to explain what we're calculating?

Comment (rswarbrick) [reply util/dvsim/Scheduler.py:377]: Brill, thanks!

**Review by rswarbrick — APPROVED** [review]
Looks good to me!

**Pull request merged**
